cli/src/commands/publish.js

```diff
@@ -0,0 +1,275 @@
+    if (params.file) {
+        return resolveBundleFromParams(cwd, params)
+    }
+    return resolveBundleFromAppConfig(cwd)
+}
+
+const promptForConfig = async params => {
+    const apiKey = params.apikey || process.env.D2_APP_HUB_API_KEY
+    if (process.env.CI && !apiKey) {
+        reporter.error('Prompt disabled in CI mode - missing apikey parameter.')
+        process.exit(1)
+    }
+
+    const responses = await inquirer.prompt([
+        {
+            type: 'input',
+            name: 'apikey',
+            message: 'App Hub API-key',
+            when: () => !params.apikey && !process.env.D2_APP_HUB_API_KEY,
+        },
+        {
+            type: 'input',
+            name: 'appId',
+            message: 'App Hub App-id',
+            when: () => params.file && !params.appId,
+        },
+        {
+            type: 'input',
+            name: 'fileVersion',
+            message: 'Version of the app',
+            when: () => params.file && !params.fileVersion,
+        },
+        {
+            type: 'input',
+            name: 'minDHIS2Version',
+            message: 'Minimum DHIS2 version supported',
+            when: () => params.file && !params.minDHIS2Version,
+            validate: v =>
+                isValidServerVersion(v) ? true : 'Invalid server version',
+        },
+        {
+            type: 'input',
+            name: 'maxDHIS2Version',
+            message: 'Maximum DHIS2 version supported',
+            when: () => params.file && !params.maxDHIS2Version,
+            validate: v =>
+                !v || isValidServerVersion(v) ? true : 'Invalid server version',
+        },
+    ])
+
+    return {
+        ...params,
+        ...responses,
+    }
+}
+
+const handler = async ({ cwd = process.cwd(), ...params }) => {
+    const publishConfig = await promptForConfig(params)
+
+    const appBundle = resolveBundle(cwd, publishConfig)
+    const uploadAppUrl = constructUploadUrl(appBundle.id)
+
+    const resolvedApiKey =
+        publishConfig.apikey || process.env.D2_APP_HUB_API_KEY
+    const client = createClient({
+        baseUrl: publishConfig.baseUrl,
+        headers: {
+            'x-api-key': resolvedApiKey,
+        },
+    })
+
+    const versionData = {
+        version: appBundle.version,
+        minDhisVersion: appBundle.minDHIS2Version,
+        maxDhisVersion: appBundle.maxDHIS2Version || '',
+        channel: publishConfig.channel,
+    }
+
+    const formData = new FormData()
+    formData.append('file', fs.createReadStream(appBundle.path))
+    formData.append('version', JSON.stringify(versionData))
+
+    try {
+        reporter.print(
+            `Uploading app bundle to ${publishConfig.baseUrl + uploadAppUrl}`
+        )
+        reporter.debug('Upload with version data', versionData)
+
+        await client.post(uploadAppUrl, formData, {
+            headers: formData.getHeaders(),
+            timeout: 30000, // Ensure we have enough time to upload a large zip file
```

Comment:
```suggestion
            timeout: 300000, // Ensure we have enough time to upload a large zip file
```

For me on a mildy-fast internet connection I was sometimes having timeouts uploading an app, let's make this nice and long (5 minutes)

Reply:
We should also ensure the app hub backend has a long timeout for that endpoint :thinking:

Reply:
@mediremi agreed, do you know what the timeout currently is?  It's at least more than 30 seconds (the previous value) since I was able to upload after bumping the client timeout

Reply:
As far as I can see we have not specified it anywhere, and the [docs](https://hapi.dev/api/?v=20.1.2#-routeoptionspayloadtimeout) says its a default value of 10 seconds, but it's obviously not? Maybe it's different for streamed payloads.

Reply:
The timeout on our load balancer in AWS is 60 seconds.

Reply:
Since we support bundles of up to 20MB maybe we should try to increase that to ~5mins?

Reply:
Sounds good to me. The docs on how to do so is here: https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/config-idle-timeout.html

We can do it using the AWS GUI as long as we don't destroy/recreate our environments that often.

We should document the AWS setup at some point as well so we know what customization we apply.

Reply:
@varl  that is different than timeout right? The timeout is for how long a request is allowed to take when uploading something AWS describes that timeout as `idle`: " If no data has been sent or received by the time that the idle timeout period elapses, the load balancer closes the connection." Data is still being sent (hopefully) when uploading takes longer than that, so to me that doesn't seem to be the same thing as we're talking about here?

Reply:
I'm not sure, you may be correct.

> For each request that a client makes through a Classic Load Balancer, the load balancer maintains two connections. The front-end connection is between the client and the load balancer. The back-end connection is between the load balancer and a registered EC2 instance. The load balancer has a configured idle timeout period that applies to its connections.

This also stood out to me as relevant:

>To ensure that lengthy operations such as file uploads have time to complete, send at least 1 byte of data before each idle timeout period elapses, and increase the length of the idle timeout period as needed.

The way I interpret the docs is that if a client uploads a file, it is sent to the load balancer, and when it's complete the load balancer sends it to the backend. It implies that the data isn't just streamed through one connection to another.

If that's true, then the docs makes sense to me since if it takes longer to upload the file to the load balancer than it maintains the connection to the backend, it would close the connection it has to the backend before the file upload completes.

I may be misunderstanding the docs, and I cannot find other docs explaining it clearly. We pay for the Load Balancer on a "per GB transferred" basis too, but I don't know if that means anything. :shrug: